src/script/common/c_content.cpp

```diff
@@ -1162,43 +1162,27 @@ NodeBox read_nodebox(lua_State *L, int index)
 }
 
 /******************************************************************************/
-MapNode readnode(lua_State *L, int index, const NodeDefManager *ndef)
+MapNode readnode(lua_State *L, int index)
 {
-	lua_getfield(L, index, "name");
-	if (!lua_isstring(L, -1))
-		throw LuaError("Node name is not set or is not a string!");
-	std::string name = lua_tostring(L, -1);
-	lua_pop(L, 1);
-
-	u8 param1 = 0;
-	lua_getfield(L, index, "param1");
-	if (!lua_isnil(L, -1))
-		param1 = lua_tonumber(L, -1);
-	lua_pop(L, 1);
-
-	u8 param2 = 0;
-	lua_getfield(L, index, "param2");
-	if (!lua_isnil(L, -1))
-		param2 = lua_tonumber(L, -1);
-	lua_pop(L, 1);
-
-	content_t id = CONTENT_IGNORE;
-	if (!ndef->getId(name, id))
-		throw LuaError("\"" + name + "\" is not a registered node!");
-
-	return {id, param1, param2};
+	lua_pushvalue(L, index);
+	lua_rawgeti(L, LUA_REGISTRYINDEX, CUSTOM_RIDX_READ_NODE);
+	lua_insert(L, -2);
+	lua_call(L, 1, 3);
+	content_t content = lua_tointeger(L, -3);
+	u8 param1 = lua_tointeger(L, -2);
+	u8 param2 = lua_tointeger(L, -1);
+	lua_pop(L, 3);
+	return MapNode(content, param1, param2);
```

Comment:
What if `content` is not a valid content id (or if the name is not a registered node)?
You need to check the return value for nil, (and do a sanity check on the int to avoid buffer-overflows when looking up the nodedef later).

Reply:
Invalid names/IDs are handled by the metatable of `name2content`. If a given name is invalid, an error will be thrown in the `__index` metamethod.

Reply:
Oh, I forgot about the metamethod.
This should be fine then, as long as `name2content` can never contain invalid entries.

Reply:
It can if mods mess with `minetest.registered_nodes`, but I doubt this will be a security issue, since invalid content IDs must be handled anyway.

Reply:
I meant invalid in such a way that you get an invalid id, that is, an id for which no nodedef exists. I don't think this is possible.

Anyway, `NodeDefManager::get` checks the content id, so you're right, invalid ids are handled anyway.

Then there's no issue here. :-)